core/trino-main/src/main/java/io/trino/operator/aggregation/arrayagg/FlatArrayBuilder.java

```diff
@@ -209,6 +209,10 @@ public void writeAll(BlockBuilder blockBuilder)
             }
         }
         int recordsInOpenGroup = ((int) size) & RECORDS_PER_GROUP_MASK;
+        if (size != 0 && recordsInOpenGroup == 0) {
+            recordsInOpenGroup = RECORDS_PER_GROUP;
+        }
```

Comment:
Maybe we can do something more direct:
```suggestion
        int recordsInOpenGroup = toIntExact(size - ((long) closedRecordGroups.size() * RECORDS_PER_GROUP));
```

Reply:
Updated. PTAL.